### PR TITLE
RAIL-11 Support for /fare/:id route

### PR DESCRIPTION
**Pull request by Dyktus**

### CODE REVIEW

File: spec/FareToken/FareToken.raml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+get:
+  description: Return fare token information
+  responses:
+    200:
+      body:
+        application/json:
+          example: !include FareTokenResponse.json
+          schema: !include FareOptionResponse.jschema
```
Comment (orzeuek) [inline spec/FareToken/FareToken.raml]: FareOptionResponse ? are you sure :)

File: src/controller/FareTokenController.ts [hunk 2]
```diff
@@ -0,0 +1,19 @@
+import { autobind } from 'core-decorators';
+import { IRouterContext } from 'koa-router';
+import { FareTokenCache } from '../service/PersistanceService';
+import { FareTokenView } from '../views/FareTokenView';
+
+@autobind
+export class FareTokenController {
```
Comment (orzeuek) [inline src/controller/FareTokenController.ts]: Please change all FareToken* to just Fare* :)

File: src/service/Container.ts [hunk 3]
```diff
@@ -219,11 +233,22 @@ export class Container {
...
-    public async linkResolver() : Promise<LinkServiceResolver> {
+    public async linkResolver(): Promise<LinkServiceResolver> {
         const linkService = httpClientFactoryProducer(this.config.service.linkService);
         return new LinkServiceResolver(linkService);
     }
 
+    @memoize
+    public async getFareTokenCache() {
+        return new FareTokenCache(
+            new MemcacheCache(
```
Comment (orzeuek) [inline src/service/Container.ts]: please use this. getMemcache()

File: src/service/PersistanceService.ts [hunk 4]
```diff
@@ -53,6 +54,30 @@ export class MemcacheCache implements Cache {
...
 }
 
+export class FareTokenCache {
+    constructor(private readonly cache: Cache) {
+    }
+
+    public async persistFareToken(data: object) {
+        const fareRegEx = new RegExp(/^\/fare\/.*/);
+        for (let link in data) {
+            if (!fareRegEx.test(link)) {
```
Comment (orzeuek) [inline src/service/PersistanceService.ts]: If I understand correctly, link is a key. data would look like:
{
"adults": 1,
"children": 0,
"fare": "/fare/5230-9328-00000-SSS-000-1-9F"
}
so we should test value (data[link]) against regex, not the key.

Comment (orzeuek) [reply src/service/PersistanceService.ts]: more: please write at least one unit test for that, using this: 
{
"/fare-option/4c9da840a1a8415facf5d7a889f5f822c0670560": {
"totalPrice": 15730,
"fares": [
{
"adults": 1,
"children": 0,
"fare": "/fare/5230-9328-00000-SSS-000-1-9F"
}
]
}
}

as a test input.

File: src/service/PersistanceService.ts [hunk 5]
```diff
@@ -53,6 +54,30 @@ export class MemcacheCache implements Cache {
     }
 }
 
+export class FareTokenCache {
+    constructor(private readonly cache: Cache) {
+    }
+
+    public async persistFareToken(data: object) {
```
Comment (orzeuek) [inline src/service/PersistanceService.ts]: I would recommend write at least one unit test for that :)

File: src/service/PersistanceService.ts [hunk 6]
```diff
@@ -67,7 +67,7 @@ export class FareTokenCache {
             const stringifiedFareToken = JSON.stringify(
                 data[link]
             );
-            return this.cache.set(link, stringifiedFareToken).catch((err) => {
+            return await this.cache.set(link, stringifiedFareToken).catch((err) => {
```
Comment (orzeuek) [inline src/service/PersistanceService.ts]: do we need that .catch() ?  if it just propagate error I think it's useless

Comment (Dyktus) [reply src/service/PersistanceService.ts]: You're right :+1:

**Pull request merged**